### DISCUSSION TITLE
asynchronous stack traces

### DISCUSSION THREAD

**Discussion by pnf** [async-profiler/async-profiler#575] [Ideas]
My company makes extensive of asynchronous (I'll spell out the full word to avoid confusion with the the async in `AsyncGetCallTrace`) programming - essentially futures on lightweight threads, but our own specialized implementation - so, for debugging and profiling purposes, we are interested in a triply combined stack: native + java + async chain.  On the debugging side, we've been able to extend IntelliJ's async stack feature (https://blog.jetbrains.com/idea/2017/02/intellij-idea-2017-1-eap-extends-debugger-with-async-stacktraces/), but profiling currently requires a mix of different tools.  
I recently prototyped (read "hacked together") a version of async-profiler to produce these triple stacks.   It works by maintaining a thread-local `jmethodId`  corresponding to the currently executing `run` method.  Then in `Profiler::recordSample`, we look for a match, and then insert the full asynchronous invocation stack (in the form of `jmethodId`s for `run` methods), with some new `BCI_ASYNC` /`FRAME_ASYNC` so it can be colored differently in `flameGraph.cpp`.  
This works and is in fact quite cool, but it's probably not the most efficient implementation, as I was compromising to avoid extensive changes to `async-profiler` that would be painful to merge in to every release.  I would like to devise an API that you would be comfortable with merging into your master branch, but would provide appropriate hooks for us and others to capture asynchronous call stacks.  Additionally, of course, I would love to hear your thoughts and suggestions on the broad problem of profiling asynchronous code.

**Comment by apangin** [comment]
It's not the first time a question of asynchronous stack traces arises in the context of profiling. I totally agree about the usefulness of this feature, and it was actually on my list of mid-term goals. I haven't yet researched the topic though. Probably, asynchronous stack traces can be implemented on top of the other related feature - a [thread-local profiling context](https://github.com/jvm-profiling-tools/async-profiler/pull/576#issuecomment-1092314951).

**Comment by pnf** [comment]
I was thinking that it was the reverse: the profiling context is a special case of async stacks where the stacks are only one deep. 
In any case, I don’t think it’s going to be possible to provide a general implementation that will work for everyone without some sort of API for the “user” to provide the additional information. The least intrusive such API I could imagine was a callback that simply provided a vector of jmethodIDs to stitch into the java stack - then it’s entirely up to us to determine how to define and maintain this stack.  In a slightly more complex version, its a vector of arbitrary longs for whose names you invoke another callback. 


Le Jeu, 7 avr., 2022, à 20:15, Andrei Pangin a écrit:

**Comment by dimkaz25** [comment]
Hi, I don’t think that these local context is a full solution. I think it needs to be bidirectional, meaning async-profiler would ‘mark’ the context as ‘was-sampled’ this way ‘user’ code can actually write the contexts relationships.

The scenario I am trying to cover is when async closures (whatever the implementation is) are fast enough that large number of them won’t get sampled. If the ‘sample’ is detected at the completion of a given async block the rest of the stack for example would be recorded

It also would allow for more complicated cases when async closures form a graph of computation and not just trees

You already have an api that allow to ‘say’ which thread to sample for example. Direct memory can be shared between c++ and Java. So this sampling and bidirectional context exchange can be done by simple updated variable. Granted at least on one side it would be thread local


Thanks 
I hope it makes sense

**Comment by pnf** [comment]
What I mean is something like this, illustrated with scala async/await:
```
  def expensive1 = burnCpu(1000, 1.1)
  def expensive2 = burnCpu(500, 1.2)
  def foo: Future[Double] = async { expensive1 }
  def bar: Future[Double] = async { await(foo) + expensive2 + expensive1  }
  def boo: Future[Double] = async { await(bar) * 2.0  }
  val myCalc = async { await(foo)+await(bar)  }

  println(ap.execute("start,event=cpu,interval=5m"))
  println(Await.result(myCalc, Duration.Inf))
  println(ap.execute("stop,file=boffo.html"))

  // Burn cpu in a way that shows a c stack
  def burnCpu(ms: Long, x0: Double): Double = {
    val t = System.nanoTime() + ms * 1000 * 1000
    var x = x0
    while(System.nanoTime() < t) { x = Math.sin(Math.pow(x,1.1)) + 1.0  }
    x
  }
```
resulting in a flame graph like
![await](https://user-images.githubusercontent.com/1265645/170882920-e1a4d537-67fc-4e46-b093-6cfce89aca47.png)

The above is generated with this POC in [async-profiler](https://github.com/pnf/async-profiler/pull/1), and the working demo code is [here](https://github.com/pnf/asyncstacks).
